File: src/duJavaAPI/Receiver.java
/*    
    Receiver.java 
    Copyright (C) 2021 Stephane Boivin (Discord: Nmare418#6397)
    
    This file is part of "DU lua sandbox API".

    "DU lua sandbox API" is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    "DU lua sandbox API" is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with "DU lua sandbox API".  If not, see <https://www.gnu.org/licenses/>.
*/
package duJavaAPI;

import java.awt.Color;
import java.util.ArrayList;

import javax.swing.JPanel;
import javax.swing.border.LineBorder;

import sandbox.execWindow;

import javax.swing.JLabel;
import javax.swing.ImageIcon;

public class Receiver extends BaseElement {
    int sizeX = 200;
    int sizeY = 100;
	public int status = 0;
	private int range;

	public Receiver(int pid, String pname, int px, int py, int prange, boolean pverboseJava) {
		
		name = pname;
		x = px;
		y = py;
		id = pid;
        range = prange;
		
		verboseJava = pverboseJava;		
		
		// create panel
		panel = new JPanel();		
		panel.setLayout(null);
		panel.setBorder(LineBorder.createBlackLineBorder());
		panel.setBackground(Color.black);		
		panel.setBounds(x, y, sizeX, sizeY);
		
		// white frame
		JPanel lblPicWhite = new JPanel();		
	 	lblPicWhite.setBounds(1, 1, sizeX-3, sizeY-3);
	 	lblPicWhite.setBackground(Color.white);
		panel.add(lblPicWhite, 1, 0);

		// picture
	 	JLabel lblPic = new JLabel();
	 	setPicture(lblPic, "src/pictures/elements/Receiver.png", 8, 5, 55, 71);
		panel.add(lblPic, 1, 0);
		
		// stats
		CreateStatPanel(name+" ("+id+")", sizeX, sizeY);		
		AddtoStatPanel("Range:", range+"m");
		panel.add(stats, 1, 0);
		
		// lua require and interface 
		urlAPI = "src/duElementAPI/receiver.lua";
        luaCall = pname+" = createInterfaceReceiver(\""+pid+"\", \""+pname+"\")";
		
        if(verboseJava) System.out.println("[JAVA] Door "+pid+".name="+pname+" created");	 		
	}

	@Override
	public ArrayList<String[]> export() {	       
	   ArrayList<String[]> listScript = new ArrayList<String[]>();
	   String sign;
	   String args;
	   String slotKey;

		for(String channel : Script.keySet()) {
		      if(Script.get(channel) == null) continue;
		      sign = "receive(channel,message)";
		      args = "[{\"value\":\"channel\"},{\"value\":\"message\"}]";
		      slotKey = Integer.toString(id);
		      listScript.add(new String[]{Script.get(channel), args, sign, slotKey}); 			
	   }		
	   	   	   
	   return listScript;
	}
	
	@Override
	public void update(String[] param) {
		// {State,name}

		// System.out.println("[JAVA] Door " + param[1] + " state set to  " + param[0]); 
	}
}

File: src/duElementAPI/receiver.lua
--[[    
    receiver.lua 
    Copyright (C) 2020 Stephane Boivin (Discord: Nmare418#6397)
    
    This file is part of "DU lua sandbox API".

    "DU lua sandbox API" is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    "DU lua sandbox API" is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with "DU lua sandbox API".  If not, see <https://www.gnu.org/licenses/>.
]]
Ireceiver = {id, name}

function Ireceiver:new(o)
  o = o or {}
  setmetatable(o, self)
  self.__index = self

  o.maxHitPoints = 2000
  o.hitPoints = 2000
  o.mass = 28.67325
  o.class = 'ReceiverUnit'
  o.integrity = 100

  o.receiver = Receiver:new{id = o.id, name = o.name}
  o.generic = genericElement:new{id = o.id, maxHitPoints = o.maxHitPoints, hitPoints = o.hitPoints, mass = o.mass, class = o.class, integrity = o.integrity }

  return o
end

function createInterfaceReceiver(id, name)

  local obj = Ireceiver:new{id=id,name=name}
 --  obj.getItemsMass = function() return obj.receiver:getItemsMass() end
  obj.getRange = function() return obj.receiver:getRange() end
  obj.RUN = function(pscript) obj.receiver:RUN(pscript) end;

  obj.hide = function() obj.generic:hide() end
  obj.show = function() obj.generic:show() end
  obj.getData = function() obj.generic:getData() end
  obj.getId = function() return obj.generic:getId() end
  obj.getMass = function() return obj.generic:getMass() end
  obj.getElementClass = function() return obj.generic:getElementClass() end
  obj.getIntegrity = function() return obj.generic:getIntegrity() end
  obj.getHitPoints = function() return obj.generic:getHitPoints() end
  obj.getMaxHitPoints = function() return obj.generic:getMaxHitPoints() end
  obj._setMass = function(m) obj.generic:setMass(m) end
  obj._setMaxHitPoints = function(mhp) obj.generic:setMaxHitPoints(mhp) end
  obj._setHitPoints = function(hp) obj.generic:setHitPoints(hp) end
  obj._setIntegrity = function(int) obj.generic:setIntegrity(int) end
   
  return obj
end

-- State 0 = closed
-- State 1 = open
Receiver = {id, name = "Receiver", range = 100}

function Receiver:new(o)
  o = o or {}
  setmetatable(o, self)
  self.__index = self

  range = 10
  
  if _verboseLUA == 1 then print("[LUA] Receiver ["..o.name.."] created with the id ["..o.id.."]") end

  return o
end

function Receiver:getRange()
  return self.range
end

